server/store/service.go

```diff
@@ -70,3 +65,19 @@ func (s *Service) hashkey(globalPrefix, namespace, prefix, id string) string {
 
 	return key
 }
+
+func hashkey(globalPrefix, namespace, prefix, id string) string {
+	namespacePrefixHash := make([]byte, 20)
+	sha3.ShakeSum128(namespacePrefixHash, []byte(namespace+prefix))
```

Comment:
My suggestion on the Jira ticket was to not hash the namespace at all. Is that not doable? I guess it comes down to the OAuth2 APIs namespacing by app ID instead of bot ID, but since all apps have a bot associated with them anyway, couldn't we use bot IDs for OAuth2, too?

The 20-byte SHAKE128 hash *should* be enough to protect namespaces against collision attacks, but in case there's ever a weakness discovered in the algorithm, we need to have a path to upgrade it. My concern is that if we don't store the unhashed keys anywhere, upgrading to a new algorithm may not be possible.

Reply:
Trying to figure it out. 
- The limit is actually 50
- The AppID is not limited in length, the bot user ID is 28. = 28
- The global prefix - need at least 1 leading byte to differentiate from all other keys (*can prefix all other keys instead) = 0
- no separator needed if using fixed 28 chars for namespace = 0
- leaves 22 bytes for app-specific prefix (type, list all keys by type) and ID
- ID is 20 bytes (16 byte hash, ascii 85)
- Apps can have a 2-letter prefixes (clear text)

Not perfect but WFM. Implementing now.

Reply:
@levb if you're still short on space with the 26-character IDs, you can also zbase32-decode them and re-encode using ASCII85. That should free up another 6 characters.

Reply:
1/5 the limit would feel artificial to a developer, but for most apps even a 1-letter prefix "should be enough!"

Reply:
FWIW, I do not consider this KV store implementation a long-term solution.

Reply:
Ended up implementing slightly differently, with a 2-char global prefix/namespace and a 2-char app-specific one.